tests/test_grad.py

```diff
@@ -0,0 +1,196 @@
+
+            test = GradTester(fn=fn,
+                              gfn=bprop_fn,
+                              args=args,
+                              argnames=argnames,
+                              outnames=[fn.__name__])
+            results = test.compare()
+
+            for arg, d in results.items():
+                if not d['match']:
+                    print(f'Argument {arg}:')
+                    print(f'\tFinite differences: {d["difference"]}')
+                    print(f'\tGradient output:    {d["exact"]}')
+                    fail(f'Mismatch in gradients for {arg} (see stdout)')
+
+        m = mark.parametrize('args', list(tests))(test)
+        m.__orig__ = fn
+        return m
+    return decorate
+
+
+@grad_test((13, 14))
+def test_null(x, y):
+    """Test null gradient."""
+    return 10 + 28 / 43
+
+
+@grad_test((3, 4))
+def test_tuple(x, y):
+    """Test multiple outputs."""
+    return (x + y, x - y, x * y, x / y)
+
+
+@grad_test((3, 4, 5))
+def test_expression(x, y, z):
+    """Test a more complex expression."""
+    return x * y + y / z
+
+
+@grad_test((3, 4, 5))
+def test_variables(x, y, z):
+    """Test an expression with temporary variables."""
+    a = x * y
+    b = y * a
+    c = a + b
+    return c / z
+
+
+@grad_test((3, 4, 5))
+def test_shadowing(x, y, z):
+    """Test an expression where variables are shadowed."""
+    x = x * y
+    y = y * z
+    z = x * z
+    return x + y + z
+
+
+@grad_test((3,))
+def test_constant(x):
+    """Test the use of a literal in the expression."""
+    return 18 * x
+
+
+@grad_test((3,))
+def test_dup_args_in_call(x):
+    """The naive gradient update rule fails when a function's arguments
+    contain the same variable more than once."""
+    return x * x
+
+
+@grad_test((3,))
+def test_quadruple_args_in_call(x):
+    """Test that duplicated arguments still cause no problem even if
+    there are four of them."""
+    def g(a, b, c, d):
+        return a * b * c * d
+    return g(x, x, x, x)
+
+
+@grad_test((4, 5))
+def test_simple_closure(a, b):
+    """Test some trivial closures."""
+    def f():
+        return a + 1
+
+    def g():
+        return b + 2
+    return f() * g()
+
+
+@grad_test((4,))
+def test_closure(a):
+    """This is the closure test in the paper."""
+    def x1(b):
+
+        def x4(c):
+            return b
+        return x4
+    x2 = x1(a)
+    x3 = x2(1)
+    return x3
+
+
+# TODO: test when a == b (finite diff won't return anything sensible)
+@grad_test((4, 5), (68, -4))
+def test_if(a, b):
+    # This is max, but what this is really testing is the most basic
+    # if statement, so I prefer to name the test 'test_if'
+    if a > b:
+        return a * b
+    else:
+        return b + a
+
+
+@grad_test((4, 5, 2), (7, 3, 1))
+def test_while(x, y, z):
+    rval = 0
+    # Cannot compare to 0 or finite diff is unstable
+    while x > -0.1:
+        rval += y
+        x -= z
+    return rval
+
+
+# @grad_test((2,), (4,), (8,))
```

Comment:
`for` is not supported by the parser and there is no `range` primitive. The test was in master, though, so I pasted it anyway and commented it out for later.

Reply:
OK, then I think we need a more formal roadmap on what needs to be ported from master, since I did not realize `for` had not been ported in the "clean" parser yet..